docs/architecture/adr-047-handling-evidence-from-light-client.md

```diff
@@ -19,161 +20,200 @@ the light client detector receives two conflicting headers
-ultimately committed. Certain fields may also be variable, for instance the
-`LastCommitHash` and the `Time` may depend on which votes the proposer includes.
-Thus, the header must be explicitly checked for invalid data.
-
-For the attack to succeed, VC must sign a header that changes the validator set
-to consist of something they control. Without doing this, they can not
-otherwise attack the light client, since the client verifies commits according
-to validator sets. Thus, it should be sufficient to check only that
-`ValidatorsHash` and `NextValidatorsHash` are correct with respect to the
-header that was committed at the corresponding height.
-
-That said, if the attack is conducted by +2/3 of the validator set, they don't
-need to make an invalid change to the validator set, since they already control
-it. Instead they would make invalid changes to the `AppHash`, or possibly other
-fields. In order to punish them, then, we would have to check all header
-fields.
-
-Note some header fields require the block itself to verify, which the light
-client, by definition, does not possess, so it may not be possible to check
-these fields. For now, then, `LunaticValidatorEvidence` must be checked against
-all header fields which are a function of the application at previous blocks.
-This includes `ValidatorsHash`, `NextValidatorsHash`, `ConsensusHash`,
-`AppHash`, and `LastResultsHash`. These should all match what's in the header
-for the block that was actually committed at the corresponding height, and
-should thus be easy to check.
-
-`InvalidHeaderField` contains the invalid field name. Note it's very likely
-that multiple fields diverge, but it's faster to check just one. This field
-MUST NOT be used to determine equality of `LunaticValidatorEvidence`.
-
-### F2. Amnesia
+The light client then checks if there has been a full fork that has caused the 
+divergence.
 
-```go
-type PotentialAmnesiaEvidence struct {
-	VoteA types.Vote
-	VoteB types.Vote
-}
+```golang
+CheckForEquivocation(h1, h2 *SignedHeader) ([]Evidence, error)
+```
+
+It first differentiates between a duplicate vote attack and an amnesia attack by
+checking the round that the commit of the divergent headers happened. If they
+are the same then it checks the individual votes and produces `DuplicateVoteEvidence`.
+If it is not the same then it bundles the signed header into amnesia evidence
+(see [ADR 056](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-056-proving-amnesia-attacks.md)).
+
+For both of these evidence, the light client send the same evidence to both
+witness and primary.
+
+It is also important here to outline that neither evidence might be detected. 
+This is because the attack that caused the full fork might have happened at a 
+height that is between the common header and the divergent headers and therefore 
+it is possible that both validator sets diverged to the point where there is 
```

Comment:
not sure I understand this. can't you still find the bifurcation point and, at minimum, create an evidence?

Reply:
I think an important distinction in terms I haven't mentioned is that there are two types of bifurcation points: The skipping bifurcation point (which is the point that two nodes diverge using skipping verification) and the sequential bifurcation point (which is the point that two nodes diverge using sequential verification). Lunatic attacks (and also back to the past) are skipping only attacks hence finding the skipping bifurcation point should guarantee that we find a lunatic attack (if it has occurred). Equivocation and Amnesia however are sequential verification attacks hence there is a chance that the attack happened not at the point of skipping bifurcation but at a height in between. In this case it is possible that the light client only observes the remnants of the attack which may no longer have any evidence

Reply:
Interesting, thanks for explaining. Would it make sense to try to find a sequential bifurcation point, which may be still within the trusting period and punishable. Or do you think punishment at skipping bifurcation point is sufficient?

Reply:
Yeah I wrote this up as possible future work and I do think it's something we should do. It will require however cooperation from both the witness and primary to find this point. But if we do the we can fully capture the equivocation/amnesia attack